### PR TITLE
Fix quarentine tests affected by AIP-39

### PR DESCRIPTION
**Pull request by uranusjr**
When we did `TaskInstance`’s migration from `execution_date` to `run_id`, we forgot to check the quarentined tests. This fixes them.

Note that this does _not_ make the quarentined tests to pass (although some do pass now), only modify them to fit the test setup required by AIP-39, namely 1. `TaskInstance.execution_date` is now fetched via a `JOIN` to `DagRun`, and an associating `DagRun` is now requried for every `TaskInstance`.

This also fixes some of the tests to match reality, which I’ll comment inline.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -795,7 +795,7 @@ def _create_dagruns_for_dags(self, guard, session):
         guard.commit()
         # END: create dagruns
 
-    def _create_dag_runs(self, dag_models: Iterable[DagModel], session: Session) -> None:
+    def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -> None:
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job.py:798]: This annotation was wrong. `dag_models` is iterated over multiple times in this function, and an iterable does not satisfy that usage (an iterator can only be iterated through once).

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -954,7 +954,7 @@ def _schedule_dag_run(
             unfinished_task_instances = (
                 session.query(TI)
                 .filter(TI.dag_id == dag_run.dag_id)
-                .filter(TI.execution_date == dag_run.execution_date)
+                .filter(TI.run_id == dag_run.run_id)
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job.py:957]: Because AIP-39. This code path does not seem to be covered by any existing case?

**Review by kaxil — APPROVED** [review]

File: tests/executors/test_celery_executor.py [hunk 3]
```diff
@@ -529,7 +529,7 @@ def test_send_tasks_to_celery_hang(register_signals):
     executor = celery_executor.CeleryExecutor()
 
     task = MockTask()
-    task_tuples_to_send = [(None, None, None, None, task) for _ in range(26)]
+    task_tuples_to_send = [(None, None, None, task) for _ in range(26)]
```
Comment (uranusjr) [inline tests/executors/test_celery_executor.py:532]: `_send_tasks_to_celery` was modified to accept 4-tuples (not 5) a while ago, but this test was never updated. Not related to AIP-39.

File: tests/jobs/test_scheduler_job.py [hunk 4]
```diff
@@ -1915,37 +1916,39 @@ def test_verify_integrity_if_dag_changed(self, dag_maker):
...
-                num_runs=1, executor=executor, subdir=os.path.join(settings.DAGS_FOLDER, "no_dags.py")
-            )
+            self.scheduler_job = SchedulerJob(num_runs=1, executor=executor, subdir=os.devnull)
+            self.scheduler_job.dagbag = dag_maker.dagbag
             self.scheduler_job.heartrate = 0
-            self.scheduler_job.run()
+            # Since the DAG is not in the directory watched by scheduler job,
+            # it would've been marked as deleted and not being scheduled.
+            with mock.patch.object(DagModel, "deactivate_deleted_dags"):
+                self.scheduler_job.run()
```
Comment (uranusjr) [inline tests/jobs/test_scheduler_job.py:1951]: This test was broken after we implemented auto DAG deactivation. Not related to AIP-39.

File: tests/jobs/test_scheduler_job.py [hunk 5]
```diff
@@ -3176,40 +3177,40 @@ def test_execute_queries_count_with_harvested_dags(self, expected_query_count, d
...
+            ([10, 10, 10, 10], 10, 10, "1d", "None", "no_structure"),
+            ([10, 10, 10, 10], 10, 10, "1d", "None", "linear"),
+            ([95, 28, 28, 28], 10, 10, "1d", "@once", "no_structure"),
+            ([105, 41, 41, 41], 10, 10, "1d", "@once", "linear"),
+            ([95, 99, 99, 99], 10, 10, "1d", "30m", "no_structure"),
+            ([105, 125, 125, 125], 10, 10, "1d", "30m", "linear"),
+            ([105, 119, 119, 119], 10, 10, "1d", "30m", "binary_tree"),
+            ([105, 119, 119, 119], 10, 10, "1d", "30m", "star"),
+            ([105, 119, 119, 119], 10, 10, "1d", "30m", "grid"),
+        ],
```
Comment (uranusjr) [inline tests/jobs/test_scheduler_job.py:3213]: Query count changes in this function are not related to AIP-39 (I tracked this breakage to a few months ago).

File: tests/jobs/test_scheduler_job.py [hunk 6]
```diff
@@ -3258,25 +3269,25 @@ def test_should_mark_dummy_task_as_success(self):
         dagbag = DagBag(dag_folder=dag_file, include_examples=False, read_dags_from_db=False)
         dagbag.sync_to_db()
 
-        self.scheduler_job_job = SchedulerJob(subdir=os.devnull)
-        self.scheduler_job_job.processor_agent = mock.MagicMock()
-        dag = self.scheduler_job_job.dagbag.get_dag("test_only_dummy_tasks")
+        self.scheduler_job = SchedulerJob(subdir=os.devnull)
+        self.scheduler_job.processor_agent = mock.MagicMock()
+        dag = self.scheduler_job.dagbag.get_dag("test_only_dummy_tasks")
```
Comment (uranusjr) [inline tests/jobs/test_scheduler_job.py:3274]: Totally unrelated to AIP-39.

**Review by potiuk — APPROVED** [review]

**Comment by ashb** [comment]
Yeah, quarantined tests that don't fail a build _essentially_ don't exist :(

File: tests/jobs/test_scheduler_job.py [hunk 7]
```diff
@@ -1669,22 +1668,24 @@ def test_scheduler_keeps_scheduling_pool_full(self, dag_maker):
...
             dag_id='test_scheduler_keeps_scheduling_pool_full_d1',
             start_date=DEFAULT_DATE,
-        ) as dag_d1:
+        ):
             BashOperator(
                 task_id='test_scheduler_keeps_scheduling_pool_full_t1',
                 pool='test_scheduler_keeps_scheduling_pool_full_p1',
                 bash_command='echo hi',
             )
+        dag_d1 = dag_maker.dag
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py:1677]: I thought  the dag_maker yielded the dag, so this style of change shouldn't be necessary?

Comment (uranusjr) [reply tests/jobs/test_scheduler_job.py:1677]: For some reason if you use `dag_maker` more than once, the `as` variables are binded to the same DAG:

```python
with dag_maker(...) as dag1:
    ...

with dag_maker(...) as dag2:
    ...

assert dag1 is dag2  # True!
```

I don't understand how this could happen, maybe there's some Pytest magic involved, but this is the easiest workaround anyway so I just did it instead of going down the rabbit hole.
